library/src-bootstrapped/scala/quoted/Unliftable.scala

```diff
@@ -1,6 +1,15 @@
 package scala.quoted
 
-/** A type class for types that can be turned from a `quoted.Expr[T]` to a `T` */
+/** A type class for types that can be turned from a `quoted.Expr[T]` to a `T`.
+ *
+ *  - Converts expression containg literal values to their values:
+ *    - `'{1}` -> `1`, `'{2}` -> `2`, ...
+ *    - For all primitive types and `String`
+ *  - Converts an expression that constructs a value into its value.
+ *    - This expression must be some kind of datastructure (`Some`, `List`, `Either`, ...)
+ *    - Calls to `new X` or `X.apply` can be lifted into its value
+ *    - Arguments of constructors can be recursively unlifted
+ */
 trait Unliftable[T] {
```

Comment:
What if a user wants to implement an instance of `Unliftable` that performs evaluation to get `T` from `Expr[T]`? I don't see a principled reason to reject that usage.

Reply:
That is simply not what unlift is for. `Unliftable` is the dual of `Liftable`, it performs the opposite operation and nothing more.

It is also impossible to implement a satisfactory version of `eval` as we would need a general-purpose interpreter which is **impossible** on the JVM due to technical detail (can be done in JS and native). It would also come with an unbounded performance cost that would render the abstraction unusable. If someone wants an `eval` function, they can implement it them selfs and can manage the balance of performance vs completeness for their precise use cases.

Reply:
There seem to be two issues here. First, the name `Unliftable` is very general. If it's only intended for normal forms, maybe it's better to rename it.

Second, as it's mentioned elsewhere, some evaluations are permissible in `Unliftable`. This creates a specification problem: what's the _principle_ there to tell what's allowed and what is not. If the restriction is not based on principles, it will be questioned and violated in practice and cause problems.

Reply:
What name would you suggest?

Reply:
I would simply keep the name, remove the restriction --- this way it's consistent.

Then I would keep `Const.unapply`, which guarantees that no evaluation happens in the unlifting.

Users should use `.unlift` most of the time. However, it is not a replacement for `Const.unapply`, as
the semantics of `.unlift` depends on the imported `Unliftable` instance --- it does not provide the semantics intended by `Const.unapply`.

Reply:
Removing the restriction would break the abstraction. There would be no way to guess what `.unlift` would do. We have already tried that and saw that without a clear specification users had no idea what to expect from half-baked evaluation schemes. This is why we went back to the simple principled design where we can only unlift constructors of values.

Reply:
> .unlift depends on the imported Unliftable instance --- it does not provide the semantics intended by Const.unapply.

It does as that is the spec. It would not only if you implement a custom `Unliftable` that does not follow the spec.

Reply:
This documentation is the current spec. We should merge this and if any changes to the spec must happen it should be done based on the current one rather than on a new one with unknown consequences.